File: python/818_RaceCar.py
from unittest import TestCase
# https://leetcode.com/problems/race-car/


class RaceCar(TestCase):
    def racecar(self, target: 'int') -> 'int':
        cache = dict()

        def cal(k):
            return (1 << k) - 1

        def dp(v):
            if v not in cache:
                i = 1
                while cal(i) < v:
                    i += 1
                if cal(i) == v:
                    cache[v] = i
                else:
                    m = dp(cal(i) - v)
                    for j in range(i-1):
                        m = min(m, j + dp(v-cal(i-1)+cal(j)))
                    cache[v] = i + 1 + m
            return cache[v]
        return dp(target)

    def test1(self):
        self.assertEqual(2, self.racecar(3))

    def test2(self):
        self.assertEqual(5, self.racecar(6))

    def test3(self):
        self.assertEqual(28, self.racecar(943))

File: kotlin/src/main/kotlin/818.RaceCar.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/race-car/


class RaceCar{
  val cache = HashMap<Int, Int>()
  private fun cal(k: Int) = (1 shl k)-1
  fun racecar(target: Int): Int {
    if(target !in cache){
      var k = 1
      while(cal(k) < target) k++
      if(cal(k) == target) cache[target] = k
      else{
        var res = racecar(cal(k)-target)
        for(j in 0 until k-1)
          res = minOf(res, j+racecar(target-cal(k-1)+cal(j)))
        cache[target] = k+1+res
      }
    }
    return cache[target]!!
  }

  @Test
  fun test1(){
    assertEquals(2, racecar(3))
  }

  @Test
  fun test2(){
    assertEquals(5, racecar(6))
  }
}